File: JavaScript/BestTimetoBuyandSellStock_IV.js
const maxProfit = function (k, prices) {
  const memo = new Map();
  function best(day, transactionsRemaining, holding) {
    if (day === prices.length || transactionsRemaining === 0) return 0;

    const key = `${day}-${transactionsRemaining}-${holding}`;

    if (memo.has(key)) {
      return memo.get(key);
    }

    let profit;

    if (holding) {
      const keepHolding = best(day + 1, transactionsRemaining, true);

      const sell =
        prices[day] + best(day + 1, transactionsRemaining - 1, false);

      profit = Math.max(keepHolding, sell);
    } else {
      const skip = best(day + 1, transactionsRemaining, false);
      const buy = -prices[day] + best(day + 1, transactionsRemaining, true);

      profit = Math.max(skip, buy);
    }
    memo.set(key, profit);
    return profit;
  }

  return best(0, k, false);
};

maxProfit(2, [2, 4, 1]);

File: Python/BestTimetoBuyandSellStock_IV.py
class Solution:
    def maxProfit(self, k: int, prices: list[int]) -> int:  
        memo = dict()

        def best(day, transactionsRemaining, holding):
            if day == len(prices) or transactionsRemaining == 0: return 0

            key  = (day, transactionsRemaining, holding)

            if key in memo: return memo[key]

            profit = None

            if holding:
                keepHolding = best(day+1, transactionsRemaining, True)
                sell = prices[day]+best(day+1, transactionsRemaining-1, False)
                profit = max(keepHolding, sell)

            else:
                skip = best(day+1, transactionsRemaining, False)
                buy = -prices[day]+best(day+1, transactionsRemaining, True)

                profit = max(skip, buy) 

            memo[key] = profit

            return profit

        return best(0, k, False)